app/src/org/commcare/sync/FirebaseMessagingDataSyncer.java

```diff
@@ -0,0 +1,334 @@
+        // Abort if FCM Message data is null
+        if (fcmMessageData == null){
+            disablePendingSyncs();
+            return;
+        }
+
+        if (!CommCareApplication.isSessionActive()) {
+            //  There is no active session at the moment, proceed accordingly
+            // TODO: Decide whether to only trigger the Sync when the 'recipient' of the message logs in
+            //  or anyone, in case multiple users are sharing the same device
+            // TODO: Decide whether to check if when there is no active session, the recipient has ever
+            //  logged in the device, before scheduling a sync post login
+            HiddenPreferences.setPendingSyncRequestFromServer(true);
+            HiddenPreferences.setPendingSyncRequestFromServerTime(fcmMessageData.getCreationTime().getMillis());
+
+            return;
+        }
+        // Retrieve the current User
+        User user = CommCareApplication.instance().getSession().getLoggedInUser();
+
+        // Check if the recipient of the message matches the current logged in user
+        if (checkUserAndDomain(user, fcmMessageData.getUsername(), fcmMessageData.getDomain())) {
+
+            if (fcmMessageData.getCreationTime().getMillis() < SyncDetailCalculations.getLastSyncTime(user.getUsername())) {
+                // A sync occurred since the sync request was triggered
+                return;
+            }
+            // Attempt to trigger the sync, according to the current state of the app
+            //The current state of the app by checking which activity is in the foreground
+            switch (getAppNavigationState()) {
+                case FORM_ENTRY:
+                    informUserAboutPendingSync(CommCareActivity.currentActivity, fcmMessageData);
+                    break;
+                case HOME_SCREEN:
+                case ENTITY_SELECTION:
+                case ENTITY_DETAIL:
+                case MENU:
+                case OTHER:
+                    triggerBackgroundSync(user);
+                    break;
+            }
+        } else {
+            // Username and Domain don't match the current user OR payload data doesn't include username
+            // and/or domain - Action: no actual, just log issue, no need to inform the user
+            Logger.log(LogTypes.TYPE_FCM, "Invalid data payload");
+        }
+    }
+
+    /**
+     * At this point, all the conditions are in place to trigger the sync: there is an active session
+     * and the current user has been successfully verified. The principle here is to:
+     * 1) Assess the current state of the app and decide the appropriate course of action, options
+     * being:
+     * - Trigger a background sync and block the user from accessing features that are not safe
+     * during a sync: logouts, syncs and form entries
+     * - Schedule the sync right after the form submission
+     */
+    private void triggerBackgroundSync(User user) {
```

Comment:
this needs to go through the usual code path way i.e. we need to upload the user forms before we can download restore to avoid overriding unsent user data with server data.